generator/parse.py

```diff
@@ -106,6 +107,87 @@ def plus_ns(self):
 ###############################################################################
 
 
+###############################################################################
+@unique
+class doc_state(Enum):
+    none = 0
+    single = 1
+    multi = 2
+    template = 3
+
+
+class comment_parser(object):
+    """parse documentation in comments"""
+    def parse(self, l, ctxt):
+        # delete '%', it marks non-key words in DAAL Doxygen documentation
+        line = l.replace('%', '')
+
+        # remove internal references in DAAL Doxygen documentation
+        m = re.match(r'^(.*), \\ref .*$', line)
+        if m:
+            line = m.group(1)
+
+        # try to find the begin of algorithm template description
+        m = re.match('^ \* <a name=\"DAAL-CLASS-ALGORITHMS__.*(BATCH|ALGORITHMIMPL)\"></a>$', line)
+        if m:
+            ctxt.doc_state = doc_state.template
+            ctxt.doc = defaultdict()
+            return True
+
+        # if parsing of bach algorithm template description is in progress
+        if ctxt.doc_state is doc_state.template:
+            # try to find template param description
+            m = re.match(r'^ \* \\tparam\s+(\w+)\s+(.*)$', line)
+            if m:
+                ctxt.doc[m.group(1)] = m.group(2)
+                return True
+            # try to find end of bach algorithm template description
+            m = re.match(r'.*\*/', line)
+            if m:
+                ctxt.doc_state = doc_state.none
+                return False
+
+        # if it is still template then go to next line
+        if ctxt.doc_state is doc_state.template:
+            return True
+
+        # try to find single line comment with documentation
+        m = re.match('^.*/\*!<(.+?)\*/.*$', line)
+        if m:
+            # save the doc to context and continue with next parser
+            ctxt.doc = m.group(1)
+            ctxt.doc_state = doc_state.single
+            return False
+
+        # try to find the begin of comment with documentation
+        m = re.match('^.*/\*!<(.+?)$', line)
+        if m:
+            # save the begin of doc to context and continue with next parser
+            ctxt.doc = m.group(1)
+            ctxt.doc_state = doc_state.multi
+            return False
+
+        # if it is multiline comment and there is place to insert it
+        if ctxt.doc_state is doc_state.multi and ctxt.doc_lambda:
+            # try to find the end of comment with documentation
+            m = re.match(r'^\s*(.+?)\*/', line)
+            if m:
+                doc = ctxt.doc_lambda()
+                doc[1] += ' ' + m.group(1)
+                ctxt.doc_state = doc_state.none
+                ctxt.doc_lambda = None
+                return False
+            # append the text in the line to documentation 
+            m = re.search(r'^\s*(.+?)$', line)
+            if m:
+                doc = ctxt.doc_lambda()
+                doc[1] += ' ' + m.group(1)
+                return True
```

Comment:
Can you avoid the code duplication by a better nesting, e.g. reset state only when at end of line but always write doc?